sky/skylet/providers/aws/node_provider.py

```diff
@@ -299,16 +299,30 @@ def create_node(self, node_config, tags, count) -> Dict[str, Any]:
                     "Values": [tags[TAG_RAY_LAUNCH_CONFIG]],
                 })
 
-            def _filter_nodes(filters, priortized_nodes=[]):
-                nodes = list(self.ec2.instances.filter(Filters=filters))
-                if priortized_nodes:
+            def _filter_nodes(filters, priortized_nodes=None):
+                remaining_nodes = list(self.ec2.instances.filter(Filters=filters))
+                if priortized_nodes is None:
+                    priortized_nodes = []
+                else:
                     priortized_node_ids = {n.id for n in priortized_nodes}
-                    nodes.sort(key=lambda n: n.id in priortized_node_ids, reverse=True)
+                    remaining_nodes = [n for n in remaining_nodes if n.id not in priortized_node_ids]
+                # This is for backward compatibility, where the uesr already has leaked
+                # stopped nodes with the different launch config before update to #1671,
+                # and the total number of the leaked nodes is greater than the number of
+                # nodes to be created. With this, we will make sure we will reuse the
+                # most recently used nodes.
+                # This can be removed in the future when we are sure all the users
+                # have updated to #1671.
+                remaining_nodes.sort(key=lambda n: n.launch_time, reverse=True)
+                nodes = priortized_nodes + remaining_nodes
                 return nodes[:count]
 
-            # SkyPilot: Try to reuse nodes with the same launch config first,
-            # then fall back to nodes with any launch config.
+            # SkyPilot: We try to use the instances with the same matching launch_config first. If
+            # there is not enough instances with matching launch_config, we then use all the
+            # instances with the same matching launch_config plus some instances with wrong
+            # launch_config.
             reuse_nodes = _filter_nodes(filters_with_launch_config)
+            assert len(reuse_nodes) <= count, len(reuse_nodes)
             if len(reuse_nodes) < count:
                 reuse_nodes = _filter_nodes(filters, priortized_nodes=reuse_nodes)
```

Comment:
Made some changes to variable naming, etc., to make it easier to read (I still feel like this func covering both cases makes it hard to read, but I don't see a better way except to expand):
```python
            def _filter_nodes_sorted_by_launch_time(filters, nodes_with_matching_launch_config=None):
                nodes = list(self.ec2.instances.filter(Filters=filters))
                if nodes_with_matching_launch_config is None:
                    nodes_with_matching_launch_config = []
                else:
                    node_ids_with_matching_launch_config = {n.id for n in nodes_with_matching_launch_config}
                    nodes = [n for n in nodes if n.id not in node_ids_with_matching_launch_config]

                # This is for backward compatibility, where the uesr already
                # has leaked stopped nodes with the different launch config
                # before update to #1671, and the total number of the leaked
                # nodes is greater than the number of nodes to be created. With
                # this, we will make sure we will reuse the most recently used
                # nodes (sorting by latest launch time first is a heuristic as
                # those node(s) are likely to be "normal").
                #
                # This can be removed in the future when we are sure all the
                # users have updated to #1671.
                nodes.sort(key=lambda n: n.launch_time, reverse=True)
                # The two lists are individually sorted by launch_time.
                nodes = nodes_with_matching_launch_config + nodes
                return nodes[:count]

            # SkyPilot: We try to use the instances with the same matching
            # launch_config first. If there is not enough instances with
            # matching launch_config, we then use all the instances with the
            # same matching launch_config plus some instances with wrong
            # launch_config.
            nodes_with_matching_launch_config = _filter_nodes_sorted_by_launch_time(filters_with_launch_config)
            # We take result[:count] in the func.
            assert len(nodes_with_matching_launch_config) <= count, len(nodes_with_matching_launch_config)

            if len(nodes_with_matching_launch_config) < count:
                reuse_nodes = _filter_nodes_sorted_by_launch_time(filters, nodes_with_matching_launch_config)
                assert len(reuse_nodes) == count, len(reuse_nodes)
            else:
                reuse_nodes = nodes_with_matching_launch_config
```

Reply:
Actually, I like the expanded version better (need to fill in comments), wdyt?

```python
nodes_matching_launch_config = list(self.ec2.instances.filter(Filters=filters_with_launch_config))
nodes_matching_launch_config.sort(key=lambda n: n.launch_time, reverse=True)

if len(nodes_matching_launch_config) >= count:
    reuse_nodes = nodes_matching_launch_config[:count]
else:
    nodes_all = list(self.ec2.instances.filter(Filters=filters))
    nodes_all.sort(key=lambda n: n.launch_time, reverse=True)

    # FIXME: actually, is this always true?
    assert len(nodes_all) >= count, len(nodes_all)

    node_ids_matching_launch_config = set(n.id for n in nodes_matching_launch_config)
    nodes_non_matching_launch_config = [n for n in nodes_all if n.id not in node_ids_matching_launch_config]

    reuse_nodes = nodes_matching_launch_config + nodes_non_matching_launch_config
    # FIXME: actually, is this always true?
    assert len(reuse_nodes) >= count, len(reuse_nodes)

    reuse_nodes = reuse_nodes[:count]
```